includes/amp-helper-functions.php

```diff
@@ -406,17 +406,33 @@ function amp_print_schemaorg_metadata() {
  *
  * @since 0.7.0
  */
-function amp_prepare_comment_post() {
-	if ( ! isset( $_GET['__amp_source_origin'] ) ) { // WPCS: CSRF ok. Beware of AMP_Theme_Support::purge_amp_query_vars().
+function amp_prepare_xhr_post() {
+	global $pagenow;
+	if ( ! isset( $_GET['__amp_source_origin'] ) || ! isset( $pagenow ) ) { // WPCS: CSRF ok. Beware of AMP_Theme_Support::purge_amp_query_vars().
 		return;
 	}
 
-	// Add amp comment hooks.
-	add_filter( 'comment_post_redirect', function() {
-		// We don't need any data, so just send a success.
-		wp_send_json_success();
-	}, PHP_INT_MAX, 2 );
-
+	if ( 'wp-comments-post.php' === $pagenow ) {
+		// This only runs on wp-comments-post.php.
+		add_filter( 'comment_post_redirect', function() {
+			// We don't need any data, so just send a success.
+			wp_send_json_success();
+		}, PHP_INT_MAX, 2 );
+	} elseif ( ! isset( $_GET['_wp_amp_action_xhr_converted'] ) ) { // WPCS: CSRF ok.
+		// submission was from a set action-xhr, implying it's being handled already.
+		return;
```

Comment:
I suggest inverting the isset check and moving the redirect filter from else to here.

Reply:
@westonruter The `wp_die_handler` and `AMP-Access-Control-Allow-Source-Origin` are fall throughs for both the comments and general submissions. moving the filter in here would make these happen for handled and unhandled. But I see what you're saying. and know how to simplify it.

Reply:
@westonruter see 83a0799 does that work?